k2/csrc/determinize.cc

```diff
@@ -0,0 +1,230 @@
+// k2/csrc/fsa_algo.cc
+
+// Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey dpove@gmail.com, Haowen Qiu qindazhu@gmail.com)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/fsa_algo.h"
+
+#include <utility>
+#include <vector>
+
+namespace k2 {
+
+
+struct DetStateElement {
+  // Element of the doubly linked list whose start/end are
+  // members 'head' and 'tail' of DetState.
+  // We can trace back the `parent` links, which will take
+  // us backward along a path in the original FSA.
+  DetStateElement *parent = nullptr;
+  int32_t arc_index;  // Index of most recent arc in path to the dest-state.
+                      // This data-structure represents a path through the FSA,
+                      // with this arc being the most recent arc on that path.
+  int32_t symbol;     // Symbol on the arc numbered `arc_index` of the input FSA
+                      // (copied here for convenience).
+
+  double weight;      // Weight from reference state to this state, along
+                      // the path taken by following the 'parent' links
+                      // (the path would have `seq_len` arcs in it).
+                      // Note: by "this state" we mean the destination-state of
+                      // the arc at `arc_index`.
+
+  // `prev` and `next` form the doubly linked list of DetStateElement
+  DetStateElement *prev = nullptr;
+  DetStateElement *next = nullptr;
+
+  // This comparator function compares the weights, but is careful in case of
+  // ties to ensure deterministic behavior.
+  bool operator < (const DetStateElement &other) const {
+    if (weight < other.weight) return true;
```

Comment:
What is the return value if `weight == other.weight` ??

Reply:
Sorry, I didn't realize I had included this code, it is not finished.  I have to get back to this.